expr/func_movingwindow.go

```diff
@@ -0,0 +1,167 @@
+package expr
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/grafana/metrictank/batch"
+	"github.com/grafana/metrictank/errors"
+	"github.com/grafana/metrictank/schema"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/consolidation"
+	"github.com/raintank/dur"
+)
+
+type FuncMovingWindow struct {
+	in           GraphiteFunc
+	windowSize   string
+	fn           string
+	xFilesFactor float64
+
+	shiftOffset uint32
+}
+
+// NewMovingWindowConstructor takes an agg string and returns a constructor function
+func NewMovingWindowConstructor(name string) func() GraphiteFunc {
+	return func() GraphiteFunc {
+		return &FuncMovingWindow{fn: name}
+	}
+}
+
+func NewMovingWindow() GraphiteFunc {
+	return &FuncMovingWindow{fn: "average", xFilesFactor: 0}
+}
+
+func (s *FuncMovingWindow) Signature() ([]Arg, []Arg) {
+	return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgString{key: "windowSize", val: &s.windowSize, validator: []Validator{IsSignedIntervalString}},
+			ArgString{key: "func", opt: true, val: &s.fn, validator: []Validator{IsConsolFunc}},
+			ArgFloat{key: "xFilesFactor", opt: true, val: &s.xFilesFactor, validator: []Validator{WithinZeroOneInclusiveInterval}},
```

Comment:
the signature is correct for movingWindow but not the other ones where the func is already known

Reply:
No sure I understand the comment here. For the children of movingWindow, we should use the `NewMovingWindowConstructor` constructor with the func passed in.

Reply:
yes but the Signature() function is what is used to validate the query (function arguments) provided by the user.
if we already set the func at construction time, we shouldn't look for an input parameter for it.
See NewAggregate() / FuncAggregate.Signature(), it's the same idea.